linode/docker-machine-driver-linode
# dep: Use GitLab-maintained docker-machine package

## 📝 Description

This pull request adds a replace statement to `go.mod` to use [GitLab's docker-machine fork](https://gitlab.com/gitlab-org/ci-cd/docker-machine) rather than [its deprecated equivalent](https://github.com/docker/machine). 

This is necessary to consume versions of Docker > v20.x.x which contain critical security fixes.

**NOTE: This has been successfully tested against v2.8.5 using the steps in the [Testing Rancher Compatibility section](#testing-rancher-compatibility).**

## ✔️ How to Test

The following test steps assume you have pulled down this PR locally.

### Unit Testing

```shell
make test
```

### Manual Testing

1. Ensure the `LINODE_TOKEN` environment variable is exported.

2. Install `docker-machine-driver-linode` locally:

```shell
make install
```

3. Download the GitLab docker-machine fork CLI locally ([see documentation](https://gitlab.com/gitlab-org/ci-cd/docker-machine/-/blob/main/docs/install-machine.md)):

```shell
curl -o docker-machine "https://gitlab-docker-machine-downloads.s3.amazonaws.com/v0.16.2-gitlab.19/docker-machine-Darwin-x86_64"
chmod +x docker-machine
```

**NOTE: Be sure to download the correct binary for your system.**

4. Provision a Linode machine:

```
# NOTE: docker-machine-driver-linode will be automatically resolved from path
./docker-machine create -d linode linode
```

5. Ensure the Linode machine provisions as expected.

### Testing Rancher Compatibility

**NOTE: This should be done in a secure dev VM with restricted access to ports 22, 80 and 443. All other ports must be open to enable connectivity between nodes and the deployed Rancher instance. See the `development-instance` tool in dx-devenv for more details**

Prerequisites:
* Docker
* Go
* Make

1. Clone and build this branch of `docker-machine-driver-linode`:

```shell
git clone https://github.com/lgarber-akamai/docker-machine-driver-linode.git --branch fix/use-gitlab-docker-machine
```

3. Build `docker-machine-driver-linode`:

```shell
# NOTE: cgo must be disabled because the Rancher container does not support glibc.
CGO_ENABLED=0 make build
```

4. Provision a Rancher container:

```shell
sudo docker run -d --restart=unless-stopped -p 80:80 -p 443:443 --privileged --name rancher rancher/rancher
```

5. Copy the built `docker-machine-driver-linode` binary into the running Rancher container:

```
docker cp ./out/docker-machine-driver-linode rancher:/opt/drivers/management-state/bin/docker-machine-driver-linode
```

6. Navigate to the VM's IP in your browser and finish the Rancher setup process.
7. Attempt to create an RKE1 (not RKE2) Linode cluster.
8. Ensure the cluster provisions as expected.
